src/vsr/replica.zig

```diff
@@ -2388,10 +2399,75 @@ pub fn Replica(
             self.state_machine.compact(commit_op_compact_callback, self.commit_prepare.?.header.op);
         }
 
-        fn commit_op_compact_callback(state_machine: *StateMachine) void {
-            const self = @fieldParentPtr(Self, "state_machine", state_machine);
+        fn commit_op_compact_callback(forest: *StateMachine.Forest) void {
+            const state_machine = @fieldParentPtr(StateMachine, "forest", forest);
+            const self = @fieldParentPtr(Replica, "state_machine", state_machine);
+            assert(self.committing);
+            assert(self.commit_callback != null);
+            assert(self.op_checkpoint == self.superblock.staging.vsr_state.commit_min);
+            assert(self.op_checkpoint == self.superblock.working.vsr_state.commit_min);
+
+            const commit = self.commit_prepare.?.header.op;
+            if ((commit + 1) % config.journal_slot_count == 0) {
+                assert((commit + 1) % config.lsm_batch_multiple == 0);
+                // Trigger the checkpoint immediately before the WAL will wrap.
+                self.state_machine.forest.checkpoint(
+                    commit_op_checkpoint_forest_callback,
+                    commit,
+                );
+            } else {
+                self.commit_op_done();
+            }
+        }
+
+        fn commit_op_checkpoint_forest_callback(forest: *StateMachine.Forest) void {
+            const state_machine = @fieldParentPtr(StateMachine, "forest", forest);
+            const self = @fieldParentPtr(Replica, "state_machine", state_machine);
+            assert(self.committing);
+            assert(self.commit_callback != null);
+            assert(self.commit_prepare.?.header.op == self.commit_min);
+            assert((self.commit_prepare.?.header.op + 1) % config.journal_slot_count == 0);
```

Comment:
I don't know if it's better but we could also get the slot for the op, to be able to compare it `journal_slot_count` in terms of the same unit?